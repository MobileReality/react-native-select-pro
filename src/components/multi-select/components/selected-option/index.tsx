import React, { ComponentPropsWithRef } from 'react';
import { Pressable, StyleSheet, Text, TextStyle, ViewStyle } from 'react-native';
import type { OptionType, Select } from '@mobile-reality/react-native-select-pro';

import { COLORS, FONT_SIZE } from '../../../../constants/styles';

type Props = {
    option: OptionType | null;
    optionWidth: number | string;
    isPlaceholder?: boolean;
    onPressRemove?: (option: OptionType) => void;
};

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    'selectControlTextStyle' | 'placeholderText'
>;

export const MultiSelectedOption = ({
    isPlaceholder = false,
    option,
    selectControlTextStyle,
    placeholderText,
    optionWidth,
    onPressRemove,
}: Props & FromSelectComponentProps) => {
    return (
        <Pressable
            accessibilityLabel={
                option ? `${(option as OptionType).label} selected` : 'Placeholder in multi-select'
            }
            onPress={() => (onPressRemove ? onPressRemove(option as OptionType) : undefined)}
            style={[
                isPlaceholder ? styles.placeholderText : styles.multiSelectionOption,
                { width: optionWidth },
            ]}>
            <Text
                numberOfLines={1}
                style={[
                    styles.text,
                    { color: option?.label ? COLORS.BLACK : COLORS.GRAY },
                    selectControlTextStyle,
                ]}>
                {option?.label || placeholderText}
            </Text>
        </Pressable>
    );
};

type Styles = {
    text: TextStyle;
    multiSelectionOption: ViewStyle;
    placeholderText: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    text: {
        fontSize: FONT_SIZE,
        textAlign: 'left',
    },
    placeholderText: {
        alignItems: 'flex-start',
        justifyContent: 'center',
        display: 'flex',
    },
    multiSelectionOption: {
        alignItems: 'flex-start',
        justifyContent: 'center',
        display: 'flex',
        borderRadius: 4,
        backgroundColor: COLORS.DISABLED,
        borderWidth: 1,
        borderColor: COLORS.BLACK,
        margin: 2,
        paddingLeft: 5,
    },
});
