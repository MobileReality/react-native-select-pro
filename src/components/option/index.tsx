import React, { ComponentPropsWithRef, forwardRef } from 'react';
import { StyleSheet, Text, TextStyle, TouchableOpacity, ViewStyle } from 'react-native';

import { COLORS, FONT_SIZE, ITEM_HEIGHT, PADDING } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers';
import type { OptionType } from '../../index';
import type { OptionsList } from '../options-list';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof OptionsList>,
    | 'optionSelectedStyle'
    | 'optionStyle'
    | 'optionTextStyle'
    | 'onSelect'
    | 'onPressOption'
    | 'OptionComponent'
>;

export type OptionProps = OptionalToRequired<FromSelectComponentProps> & {
    isSelected: boolean;
    option: OptionType;
};

export type OnChooseOption = () => void;

export const Option = forwardRef<TouchableOpacity, OptionProps>(
    (
        {
            optionSelectedStyle,
            optionStyle,
            optionTextStyle,
            isSelected,
            onPressOption,
            option,
            onSelect,
            OptionComponent,
        },
        ref,
    ) => {
        const { label } = option;

        const onChooseOption: OnChooseOption = () => {
            onPressOption(option);
            if (onSelect) {
                onSelect(option);
            }
        };

        if (OptionComponent) {
            return (
                <OptionComponent
                    isSelected={isSelected}
                    onPressOption={onChooseOption}
                    option={option}
                />
            );
        }

        return (
            <TouchableOpacity
                accessibilityLabel={`Choose ${label} option`}
                accessibilityRole={'button'}
                accessible={true}
                onPress={onChooseOption}
                ref={ref}
                style={[
                    styles.option,
                    optionStyle,
                    isSelected && [styles.selected, optionSelectedStyle],
                ]}>
                <Text numberOfLines={1} style={[styles.text, optionTextStyle]}>
                    {label}
                </Text>
            </TouchableOpacity>
        );
    },
);

type Styles = {
    option: ViewStyle;
    selected: ViewStyle;
    text: TextStyle;
};

const styles = StyleSheet.create<Styles>({
    option: {
        height: ITEM_HEIGHT,
        justifyContent: 'center',
        paddingHorizontal: PADDING,
    },
    text: {
        fontSize: FONT_SIZE,
        color: COLORS.BLACK,
        textAlign: 'left',
    },
    selected: {
        backgroundColor: COLORS.SELECTED,
    },
});
