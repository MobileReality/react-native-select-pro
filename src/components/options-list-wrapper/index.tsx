import React, { ComponentProps, ReactElement, useRef } from 'react';
import { Animated, StyleProp, View, ViewStyle } from 'react-native';

import { ANIMATION_DURATION } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers';
import type { Select } from '../../index';
import type { OptionsList } from '../options-list';

type FromSelectProps = Pick<ComponentProps<typeof Select>, 'isAnimated'>;
type FromOptionListProps = Pick<ComponentProps<typeof OptionsList>, 'isOpened'>;

type WrapperStyles = {
    wrapperStyles: StyleProp<ViewStyle>;
};

type OptionsListWrapperProps = OptionalToRequired<FromSelectProps & FromOptionListProps> & {
    children: ReactElement;
} & WrapperStyles;

export const OptionsListWrapper = ({
    children,
    isAnimated,
    isOpened,
    wrapperStyles,
}: OptionsListWrapperProps) => {
    const fadeAnimation = useRef(new Animated.Value(0)).current;

    React.useEffect(() => {
        if (isAnimated) {
            Animated.timing(fadeAnimation, {
                toValue: isOpened ? 1 : 0,
                duration: ANIMATION_DURATION,
                useNativeDriver: true,
            }).start();
        }
    }, [fadeAnimation, isOpened]);

    return isAnimated ? (
        <Animated.View
            pointerEvents={isOpened ? 'auto' : 'none'}
            style={[wrapperStyles, { opacity: fadeAnimation }]}>
            {children}
        </Animated.View>
    ) : isOpened ? (
        <View style={wrapperStyles}>{children}</View>
    ) : null;
};
