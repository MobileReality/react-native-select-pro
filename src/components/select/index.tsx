import React, {
    ForwardedRef,
    forwardRef,
    useEffect,
    useImperativeHandle,
    useReducer,
    useRef,
} from 'react';
import { StyleSheet, useWindowDimensions, View, ViewStyle } from 'react-native';

import { MAX_HEIGHT_LIST } from '../../constants/styles';
import { initialData, reducer } from '../../state/reducer';
import { Action } from '../../state/types';
import type {
    OnOutsidePress,
    OnPressOptionType,
    OnPressSelectControlType,
    OptionType,
    SelectProps,
    SelectRef,
} from '../../types';
import { OptionsList } from '../options-list';
import { SelectControl } from '../select-control';

export const Select = forwardRef((props: SelectProps, ref: ForwardedRef<SelectRef>) => {
    const {
        options,
        clearable = true,
        closeDropdownOnSelect = true,
        defaultOption,
        disabled = false,
        flatListProps,
        hideSelectControlArrow,
        noOptionsText = 'No options',
        onSelect,
        onDropdownOpened,
        onDropdownClosed,
        optionSelectedStyle,
        optionStyle,
        optionTextStyle,
        placeholderText = 'Select...',
        selectContainerStyle,
        selectControlButtonsContainerStyle,
        selectControlClearOptionButtonStyle,
        selectControlClearOptionImageStyle,
        selectControlClearOptionButtonHitSlop,
        selectControlClearOptionA11yLabel,
        selectControlOpenDropdownA11yLabel,
        selectControlCloseDropdownA11yLabel,
        selectControlDisabledStyle,
        selectControlStyle,
        selectControlTextStyle,
        optionsListStyle,
        NoOptionsComponent,
        OptionComponent,
    } = props;
    const [{ isOpened, selectedOption, optionsData, openedPosition }, dispatch] = useReducer(
        reducer,
        initialData,
    );
    const { aboveSelectControl } = openedPosition;

    const containerRef = useRef<View>(null);

    useEffect(() => {
        if (!Array.isArray(options)) {
            console.error('You must pass array in the options prop');
            return;
        }

        if (options.length > 0) {
            dispatch({ type: Action.SetOptionsData, payload: options });

            if (typeof defaultOption === 'object') {
                dispatch({ type: Action.SelectOption, payload: defaultOption });
            }
        }
    }, [options]);

    useImperativeHandle(ref, () => ({
        clear: () => {
            dispatch({ type: Action.SelectOption, payload: null });
            dispatch({ type: Action.SetOptionsData, payload: options });
        },
        open: () => {
            if (containerRef.current && !disabled) {
                dispatch({
                    type: Action.Open,
                });
                setPosition();
            }
        },
        close: () => {
            dispatch({
                type: Action.Close,
            });
        },
    }));

    const onPressOption: OnPressOptionType = (option: OptionType) => {
        if (closeDropdownOnSelect) {
            dispatch({ type: Action.Close });
        }
        dispatch({ type: Action.SelectOption, payload: option });
        dispatch({ type: Action.SetOptionsData, payload: options });
    };

    const windowDimensions = useWindowDimensions();

    const setPosition = () => {
        if (containerRef.current) {
            containerRef.current.measure((_x, _y, width, height, pageX, pageY) => {
                const fromProp = StyleSheet.flatten(optionsListStyle)?.maxHeight;
                let listHeight = MAX_HEIGHT_LIST;
                // TODO string values like percents
                if (typeof fromProp === 'number') {
                    listHeight = StyleSheet.flatten(optionsListStyle).maxHeight as number;
                }

                const isOverflow = pageY + height + listHeight > windowDimensions.height;
                dispatch({
                    type: Action.SetPosition,
                    payload: {
                        width,
                        top: isOverflow ? pageY - listHeight : pageY + height,
                        left: pageX,
                        aboveSelectControl: isOverflow,
                    },
                });
            });
        }
    };

    const onPressSelectControl: OnPressSelectControlType = () => {
        if (isOpened) {
            dispatch({
                type: Action.Close,
            });
            return;
        }
        setPosition();
        if (containerRef.current) {
            dispatch({
                type: Action.Open,
            });
        }
    };

    const onOutsidePress: OnOutsidePress = () => {
        dispatch({ type: Action.Close });
        dispatch({ type: Action.SetOptionsData, payload: options });
    };

    useEffect(() => {
        if (isOpened) {
            onDropdownOpened?.();
        } else {
            onDropdownClosed?.();
        }
    }, [isOpened]);

    return (
        <View onLayout={setPosition} style={[styles.relative, selectContainerStyle]}>
            <SelectControl
                aboveSelectControl={aboveSelectControl}
                clearable={clearable}
                disabled={disabled}
                dispatch={dispatch}
                hideSelectControlArrow={hideSelectControlArrow}
                isOpened={isOpened}
                onPressSelectControl={onPressSelectControl}
                onSelect={onSelect}
                options={options}
                placeholderText={placeholderText}
                ref={containerRef}
                selectControlButtonsContainerStyle={selectControlButtonsContainerStyle}
                selectControlClearOptionA11yLabel={selectControlClearOptionA11yLabel}
                selectControlClearOptionButtonHitSlop={selectControlClearOptionButtonHitSlop}
                selectControlClearOptionButtonStyle={selectControlClearOptionButtonStyle}
                selectControlClearOptionImageStyle={selectControlClearOptionImageStyle}
                selectControlCloseDropdownA11yLabel={selectControlCloseDropdownA11yLabel}
                selectControlDisabledStyle={selectControlDisabledStyle}
                selectControlOpenDropdownA11yLabel={selectControlOpenDropdownA11yLabel}
                selectControlStyle={selectControlStyle}
                selectControlTextStyle={selectControlTextStyle}
                selectedOption={selectedOption}
            />
            <OptionsList
                NoOptionsComponent={NoOptionsComponent}
                OptionComponent={OptionComponent}
                aboveSelectControl={aboveSelectControl}
                flatListProps={flatListProps}
                isOpened={isOpened}
                noOptionsText={noOptionsText}
                onOutsidePress={onOutsidePress}
                onPressOption={onPressOption}
                onSelect={onSelect}
                openedPosition={openedPosition}
                optionSelectedStyle={optionSelectedStyle}
                optionStyle={optionStyle}
                optionTextStyle={optionTextStyle}
                optionsData={optionsData}
                optionsListStyle={optionsListStyle}
                selectedOption={selectedOption}
            />
        </View>
    );
});

type Styles = {
    relative: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    relative: {
        position: 'relative',
    },
});
