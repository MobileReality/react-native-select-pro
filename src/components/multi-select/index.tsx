import React, { ComponentPropsWithRef } from 'react';
import { ScrollView, StyleSheet, useWindowDimensions, ViewStyle } from 'react-native';
import type { OptionType, Select } from '@mobile-reality/react-native-select-pro';
import type {
    OnPressSelectControlType,
    OnSetPosition,
} from '@mobile-reality/react-native-select-pro';

import type { OptionalToRequired } from '../../helpers';
import { parsePercentageToNumber } from '../../helpers/parsePercentageToNumber';
import type { State } from '../../state/types';
import type { DispatchType } from '../../state/types';
import { SelectInput } from '../select-input';

import { MultiSelectedOption } from './components/selected-option';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'selectControlTextStyle'
    | 'placeholderText'
    | 'selectControlStyle'
    | 'disabled'
    | 'searchPattern'
    | 'searchable'
    | 'multiSelection'
>;

type SelectControlProps = OptionalToRequired<
    FromSelectComponentProps & Pick<State, 'selectedOption'>
>;

type Props = {
    onPressRemove: (option: OptionType | null) => void;
    onPressSelectControl: OnPressSelectControlType;
} & SelectControlProps &
    Pick<State, 'isOpened' | 'selectedOption' | 'searchValue'> & {
        dispatch: DispatchType;
    } & { setPosition: OnSetPosition };

export const MultiSelect = ({
    searchable,
    selectControlTextStyle,
    selectControlStyle,
    selectedOption,
    placeholderText,
    onPressRemove,
    disabled,
    dispatch,
    isOpened,
    onPressSelectControl,
    searchPattern,
    searchValue,
    setPosition,
    multiSelection,
}: Props) => {
    const { width } = useWindowDimensions();
    const selectedOptionTyped = selectedOption as OptionType[];

    const resolveSelectedOptions = () => {
        if (!selectedOptionTyped) {
            if (searchable) {
                return null;
            }
            return (
                <MultiSelectedOption
                    isPlaceholder={true}
                    option={null}
                    optionWidth={'100%'}
                    placeholderText={placeholderText}
                    selectControlTextStyle={selectControlTextStyle}
                />
            );
        }

        const optionWidth = () => {
            const WIDTH_THRESHOLD = 100;
            const WIDTH_OFFSET = 72;
            const length = selectedOptionTyped.length;
            const initialWidth = selectControlStyle ? (selectControlStyle as ViewStyle).width : 100;
            let calculatedWidth = 100;
            if (typeof initialWidth === 'number') {
                calculatedWidth = (initialWidth - WIDTH_OFFSET) / length;
                if (calculatedWidth < WIDTH_THRESHOLD) {
                    return WIDTH_THRESHOLD;
                }
                return Math.floor(calculatedWidth);
            }
            if (typeof initialWidth === 'string') {
                const ratioToScreen = Math.floor(
                    width * (parsePercentageToNumber(initialWidth) / 100),
                );
                calculatedWidth = ratioToScreen / length;
                if (calculatedWidth - WIDTH_OFFSET < WIDTH_THRESHOLD) {
                    return WIDTH_THRESHOLD;
                }
                return calculatedWidth - WIDTH_OFFSET;
            }
            return 0;
        };

        return selectedOptionTyped.map((option: OptionType, index) => {
            return (
                <MultiSelectedOption
                    key={index}
                    onPressRemove={onPressRemove}
                    option={option}
                    optionWidth={optionWidth()}
                    placeholderText={placeholderText}
                    selectControlTextStyle={selectControlTextStyle}
                />
            );
        });
    };

    return (
        <ScrollView horizontal={true} style={styles.multiSelectionWrapper}>
            {searchable && (
                <SelectInput
                    disabled={disabled}
                    dispatch={dispatch}
                    isOpened={isOpened}
                    multiSelection={multiSelection}
                    onPressSelectControl={onPressSelectControl}
                    placeholderText={placeholderText}
                    searchPattern={searchPattern}
                    searchValue={searchValue}
                    selectedOption={selectedOption}
                    setPosition={setPosition}
                />
            )}
            {resolveSelectedOptions()}
        </ScrollView>
    );
};

type Styles = {
    multiSelectionWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    multiSelectionWrapper: {
        flex: 1,
    },
});
