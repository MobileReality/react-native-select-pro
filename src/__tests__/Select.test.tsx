import React from 'react';
import { MeasureOnSuccessCallback, Pressable, Text, View } from 'react-native';
import type { OptionComponentProps } from '@mobile-reality/react-native-select-pro';
import { fireEvent, render } from '@testing-library/react-native';

import { Select } from '../components/select';
import { SelectProvider } from '../components/select-provider';

const DATA = [
    {
        value: 'bd7acbea-c1b1-46c2-aed5-3ad53abb28ba',
        label: 'First Option',
    },
    {
        value: '3ac68afc-c605-48d3-a4f8-fbd91aa97f63',
        label: 'Second Option',
    },
    {
        value: '58694a0f-3da1-471f-bd96-145571e29d72',
        label: 'Third Option',
    },
];

const SEARCHABLE_DATA = [
    {
        value: 'test1',
        label: 'Fist test options',
    },
    {
        value: 'test2',
        label: 'Second test options',
    },
    {
        value: 'test3',
        label: 'Third test option',
    },
];

describe('Select', () => {
    it('should generate Select snapshot', () => {
        const wrapper = render(
            <SelectProvider>
                <Select options={DATA} />
            </SelectProvider>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    jest.spyOn(View.prototype, 'measure').mockImplementation(
        (fn: MeasureOnSuccessCallback): void => {
            fn(0, 0, 45, 20, 0, 0);
        },
    );

    it('should open options menu after press select control', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={DATA} />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Open a dropdown');
        fireEvent.press(open);

        const list = getByA11yLabel('Options list');
        expect(list).toBeTruthy();
    });

    /**
     * to achieve this you need to pass `clearable={true}` and should be selected option
     */
    it('should show clear selected option button', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select clearable={true} options={DATA} />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Open a dropdown');
        fireEvent.press(open);

        const optionPress = getByA11yLabel(`Choose ${DATA[0].label} option`);
        fireEvent.press(optionPress);

        const clearButton = getByA11yLabel('Clear a chosen option');
        expect(clearButton).toBeTruthy();
    });

    it('should close dropdown menu after pressed outside dropdown', () => {
        const { getByA11yLabel, queryByA11yLabel } = render(
            <SelectProvider>
                <Select options={DATA} />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Open a dropdown');
        fireEvent.press(open);

        const listWrapper = queryByA11yLabel('Options list');
        expect(listWrapper).toBeTruthy();

        const outside = getByA11yLabel('Close a dropdown from outside');
        fireEvent.press(outside);

        const listWrapper2 = queryByA11yLabel('Options list');
        expect(listWrapper2).toBeFalsy();
    });

    it('should select option with custom option component', () => {
        const MyCustomOption = ({ onPressOption, option }: OptionComponentProps) => {
            return (
                <Pressable accessibilityLabel={'Option'} onPress={onPressOption}>
                    <Text>{option.label}</Text>
                </Pressable>
            );
        };
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select
                    OptionComponent={(props) => <MyCustomOption {...props} />}
                    options={[DATA[0]]}
                />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Open a dropdown');
        fireEvent.press(open);

        const list = getByA11yLabel('Options list');
        expect(list).toBeTruthy();

        const option = getByA11yLabel('Option');
        fireEvent.press(option);
    });
});

describe('Select with searchable enabled', () => {
    it('should generate Select with searchable enabled snapshot', () => {
        const wrapper = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should open options menu after pressing Pressable in select control with searchable enabled', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Open a dropdown');
        fireEvent.press(open);

        const list = getByA11yLabel('Options list');
        expect(list).toBeTruthy();
    });

    it('should NOT open options menu after pressing Pressable or Input in select control with searchable enabled and whole select disabled', () => {
        const onOpen = jest.fn();

        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select
                    disabled={true}
                    onDropdownOpened={onOpen}
                    options={SEARCHABLE_DATA}
                    searchable={true}
                />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Open a dropdown');
        fireEvent.press(open);
        expect(onOpen).not.toHaveBeenCalled();

        const testInput = getByA11yLabel('Place text');
        fireEvent.press(testInput);
        expect(onOpen).not.toHaveBeenCalled();
    });

    it('should open options menu after pressing Input in select control with searchable enabled', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );

        const open = getByA11yLabel('Place text');
        fireEvent.press(open);

        const list = getByA11yLabel('Options list');
        expect(list).toBeTruthy();
    });

    it('should type text in Input in select control with searchable enabled', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );

        const inputData = 'Testing messing';

        const input = getByA11yLabel('Place text');
        fireEvent.changeText(input, inputData);

        expect(input.props.value).toBe(inputData);
    });

    it('should not able to type text in Input in select control with searchable enabled but disabled', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select disabled={true} options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );

        const inputData = 'Testing messing';

        const input = getByA11yLabel('Place text');
        fireEvent.changeText(input, inputData);

        expect(input.props.value).toBe('');
    });

    it('should, while searchable enabled, search by label and default pattern', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );

        const firstInputData = 'Second test options';
        const secondInputData = 'Third test option';
        const thirdInputData = 'options';

        const input = getByA11yLabel('Place text');

        fireEvent.changeText(input, firstInputData);

        const list = getByA11yLabel('Options list');

        expect(list.props.data.length).toBe(1);

        fireEvent.changeText(input, secondInputData);
        expect(list.props.data.length).toBe(1);

        fireEvent.changeText(input, thirdInputData);
        expect(list.props.data.length).toBe(2);

        const valueData = 'test1';

        fireEvent.changeText(input, valueData);
        expect(list.props.data.length).toBe(0);
    });

    it('should, while searchable enabled, search by label and other patterns', () => {
        const searchPattern = (value: string) => `^${value}`;

        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchPattern={searchPattern} searchable={true} />
            </SelectProvider>,
        );

        const firstInputData = 'Second';
        const secondInputData = 'option';
        const input = getByA11yLabel('Place text');

        fireEvent.changeText(input, firstInputData);

        const list = getByA11yLabel('Options list');

        expect(list.props.data.length).toBe(1);

        fireEvent.changeText(input, secondInputData);
        expect(list.props.data.length).toBe(0);
    });

    it('should, while searchable enabled, get back to previous value in select if clicked outside ', () => {
        const { getByA11yLabel } = render(
            <SelectProvider>
                <Select options={SEARCHABLE_DATA} searchable={true} />
            </SelectProvider>,
        );

        const inputData = 'Second';
        const nextInputData = 'option';

        const selectOptionInputData = 'Second test options';

        const input = getByA11yLabel('Place text');

        fireEvent.changeText(input, inputData);

        const list = getByA11yLabel('Options list');

        expect(list.props.data.length).toBe(1);

        const option = getByA11yLabel(`Choose ${inputData} test options option`);

        fireEvent.press(option);

        expect(input.props.value).toBe(selectOptionInputData);

        fireEvent.changeText(input, nextInputData);

        const listAgain = getByA11yLabel('Options list');
        expect(listAgain.props.data.length).toBeGreaterThan(1);
        expect(input.props.value).toBe(nextInputData);

        const outside = getByA11yLabel('Close a dropdown from outside');
        fireEvent.press(outside);

        expect(input.props.value).toBe(selectOptionInputData);
    });
});
