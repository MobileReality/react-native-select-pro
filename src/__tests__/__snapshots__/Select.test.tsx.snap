// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should generate Select snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <View
      accessibilityLabel="Open a dropdown"
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
          },
          Object {},
          undefined,
          Object {},
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
            "paddingHorizontal": 12,
            "paddingRight": 55,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontSize": 12,
                "textAlign": "left",
              },
              Object {
                "color": "#808080",
              },
              undefined,
            ]
          }
        >
          Select...
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "100%",
              "justifyContent": "center",
              "position": "absolute",
              "right": 8,
              "top": 0,
            },
            undefined,
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/icons/chevron-down.png",
            }
          }
          style={
            Array [
              Object {
                "height": 25,
                "width": 25,
                "zIndex": -1,
              },
              Object {
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Select with custom left icon should generate Select with custom left icon snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <View
      accessibilityLabel="Open a dropdown"
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
          },
          Object {},
          undefined,
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingLeft": 8,
            },
            Object {
              "height": "100%",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/__tests__/assets/search.png",
            }
          }
          style={
            Object {
              "height": 15,
              "width": 15,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
            "paddingHorizontal": 12,
            "paddingRight": 55,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontSize": 12,
                "textAlign": "left",
              },
              Object {
                "color": "#808080",
              },
              undefined,
            ]
          }
        >
          Select...
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "100%",
              "justifyContent": "center",
              "position": "absolute",
              "right": 8,
              "top": 0,
            },
            undefined,
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/icons/chevron-down.png",
            }
          }
          style={
            Array [
              Object {
                "height": 25,
                "width": 25,
                "zIndex": -1,
              },
              Object {
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Select with searchable enabled should generate Select with searchable enabled snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <View
      accessibilityLabel="Open a dropdown"
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
          },
          Object {},
          undefined,
          Object {},
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
            "paddingHorizontal": 12,
            "paddingRight": 55,
          }
        }
      >
        <TextInput
          accessibilityLabel="Place text"
          editable={true}
          onChangeText={[Function]}
          onPressIn={[Function]}
          placeholder="Select..."
          placeholderTextColor="#808080"
          style={
            Object {
              "fontSize": 12,
            }
          }
          textAlign="left"
          value=""
        />
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "100%",
              "justifyContent": "center",
              "position": "absolute",
              "right": 8,
              "top": 0,
            },
            undefined,
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/icons/chevron-down.png",
            }
          }
          style={
            Array [
              Object {
                "height": 25,
                "width": 25,
                "zIndex": -1,
              },
              Object {
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
